### PR TITLE
Cover legacy HA-compat endpoints end-to-end

### PR DESCRIPTION
**Pull request by bdraco**
## Summary
- Adds `tests/api/test_legacy.py` — 22 E2E tests driving every route in `api/legacy.py` through a real aiohttp app: `/devices`, `/json-config`, `/compile` (WS spawn), `/upload` (WS spawn). Frame shapes, key set, error codes, and the spawn-protocol edge cases (non-JSON frames, missing port, traversal rejection emitting `{event: "exit", code: 1}` rather than tearing down the WS) are pinned to match HA's `esphome-dashboard-api` contract.
- Surfaced and fixed two production bugs in `/json-config` that meant every well-formed request was returning `500`:
  - `yaml_util.load_yaml` requires a `Path` (it calls `fname.open(...)`); passing `str(config_path)` crashed with `AttributeError` and the bare `except` surfaced it as an opaque `500`.
  - ESPHome's `load_yaml` returns an `OrderedDict` whose keys are `EStr` (a `str` subclass). orjson rejects non-exact-`str` keys; switched to `web.json_response(config, dumps=json.dumps)` for this one endpoint. HA's library accepts whatever JSON we produce.
- Drops the older `tests/test_legacy_traversal.py` (consolidated into the new file).

## Test plan
- [x] `uv run pytest tests/api/test_legacy.py` — 22 passed
- [x] `uv run pytest` — full suite passes
- [x] `uv run pre-commit run --all-files` clean
- [x] Verified frame contract against upstream `esphome/dashboard/web_server.py` (`EsphomeCommandWebSocket`, `JsonConfigRequestHandler`, `ListDevicesHandler`)

### CODE REVIEW

File: tests/api/test_legacy.py [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+
+
+@pytest.fixture
+def captured_spawn(monkeypatch: pytest.MonkeyPatch) -> dict[str, Any]:
+    """Replace ``create_subprocess_exec`` with a recording fake.
+
+    Test asserts on ``captured["cmd"]`` to verify the command
+    shape (this is the upstream-parity check). Default fake
+    yields two output lines and exits 0; tests that need
+    different behaviour overwrite ``captured["proc_factory"]``.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/api/test_legacy.py]: Fixed in 83ac706 — rewrote the docstring to describe the actual override mechanism (mutate `captured["lines"]` / `captured["exit_code"]` before opening the WS).

File: tests/api/test_legacy.py [hunk 2]
```diff
@@ -0,0 +1,719 @@
...
+async def test_spawn_ws_real_subprocess_streams_output(
+    tmp_path: Path, aiohttp_client: AiohttpClient
+) -> None:
+    """Real subprocess (no mock) → output streams through, exit code passes.
+
+    True end-to-end: drives the actual ``create_subprocess_exec``
+    code path against a real shell command. Catches a regression
+    where a refactor of the streaming loop or the helper itself
+    breaks the line iteration. Uses ``echo`` so the test
+    doesn't depend on having ``esphome`` installed.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/api/test_legacy.py]: Fixed in 83ac706 — docstring now matches the implementation (`sys.executable -c '<one-liner>'`, not `echo`).

File: tests/api/test_legacy.py [hunk 3]
```diff
@@ -0,0 +1,719 @@
...
+    # Line frame.
+    assert (
+        '{"event": "line", "data":' in src
+        or "{'event': 'line'" in src
+        or ("event" in src and '"line"' in src and '"data"' in src)
+    )
+    # Exit frame on traversal-rejection (literal "code": 1).
+    assert json.dumps({"event": "exit", "code": 1}) in src or (
+        '"event": "exit"' in src and '"code"' in src
+    )
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/api/test_legacy.py]: Removed in 83ac706. Agreed it was brittle and redundant — every behavioral WS test above already pins the exact frame keys against actual frames (`assert msg == {"event": "line", "data": ...}` / `assert msg == {"event": "exit", "code": ...}`), so the source-text-matching duplicate is gone.

File: esphome_device_builder/helpers/json.py [hunk 4]
```diff
@@ -33,6 +33,26 @@ def dumps(obj: Any) -> bytes:
...
+    str``. Use this helper for the legacy ``/json-config`` endpoint
+    (and only there); the strict default still catches the more
+    common bug shape — a dict with non-string keys leaking into a
+    response — for every other call site.
+
+    Returns ``str`` so it can be passed straight to aiohttp's
+    ``web.json_response(dumps=...)`` (which expects a ``str``-
+    returning callable, like ``dumps_str``).
+    """
+    return orjson.dumps(obj, option=orjson.OPT_NON_STR_KEYS).decode()
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/helpers/json.py:57]: Renamed to `dumps_str_non_str_keys` in 6071190 to match the actual scope of `OPT_NON_STR_KEYS` (permits int, float, datetime, UUID, etc., not just str subclasses). The single caller — legacy `/json-config` — only ever feeds EStr-keyed dicts, so the broader permissiveness is academic in practice; updated the docstring to spell out the wider key set so the name no longer overpromises the guardrail.

File: tests/api/test_legacy.py [hunk 5]
```diff
@@ -0,0 +1,695 @@
...
+@pytest.fixture
+def captured_spawn(monkeypatch: pytest.MonkeyPatch) -> dict[str, Any]:
+    """Replace ``create_subprocess_exec`` with a recording fake.
+
+    Tests assert on ``captured["cmd"]`` to verify the command
+    shape (this is the upstream-parity check). The default fake
+    yields two output lines and exits 0; a test that needs
+    different behaviour mutates ``captured["lines"]`` /
+    ``captured["exit_code"]`` *before* opening the WS — the
+    fake reads them at spawn time, not at fixture-build time.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/api/test_legacy.py:369]: Already fixed in 83ac706 (same fix as the earlier thread above) — the docstring now describes the actual override mechanism (mutate `captured["lines"]` / `captured["exit_code"]` before opening the WS). This appears to be a duplicate flagged on the next review pass.
